src/Hedgehog/Extras/Test/New/Monad/PropertyIO.hs

```diff
@@ -0,0 +1,114 @@
+import Data.Generics.Product.Any
+import Data.Maybe
+import Data.Monoid
+import HaskellWorks.Control.Monad
+import HaskellWorks.Prelude
+import Hedgehog hiding (forAll)
+import Hedgehog qualified as H
+import Hedgehog.Extras.Test.MonadAssertion
+import Hedgehog.Internal.Property qualified as H
+import Lens.Micro
+import Test.Tasty.Discover
+import Test.Tasty.Hedgehog (testProperty)
+
+newtype PropertyEnv = PropertyEnv { unPropertyEnv :: TMVar (PropertyT IO ()) }
+
+newtype PropertyIO a = PropertyIO { runPropertIO :: ExceptT H.Failure (ResourceT (ReaderT PropertyEnv IO)) a }
+  deriving newtype (Applicative)
+  deriving newtype (Functor)
+  deriving newtype (Monad)
+  deriving newtype (MonadCatch)
+  deriving newtype (MonadError H.Failure)
+  deriving newtype (MonadFail)
+  deriving newtype (MonadFix)
+  deriving newtype (MonadIO)
+  deriving newtype (MonadMask)
+  deriving newtype (MonadReader PropertyEnv)
+  deriving newtype (MonadResource)
+  deriving newtype (MonadThrow)
+  deriving newtype (PrimMonad)
+
+instance MonadTest PropertyIO where
+  liftTest f = do
+    PropertyEnv tmvAction <- ask @PropertyEnv
+    mvA <- liftIO STM.newEmptyTMVarIO
+    liftIO $ STM.atomically $ putTMVar tmvAction $
+      tryExceptAssertion (liftTest f) >>= liftIO . STM.atomically . STM.putTMVar mvA
+    testResult <- liftIO $ STM.atomically $ STM.takeTMVar mvA
+    getTestResult testResult
+
+instance Tasty (PropertyIO ()) where
+  tasty info = pure . testProperty testName . H.withTests 1 . H.property . propertyIOToPropertyT
+    where testName = fromMaybe "" $ getLast (info ^. the @"name")
+
+propertyIOToPropertyT :: ()
+  => PropertyIO ()
+  -> H.PropertyT IO ()
+propertyIOToPropertyT f = do
+  tvResult <- liftIO STM.newEmptyTMVarIO
+  tvAction <- liftIO STM.newEmptyTMVarIO
+  CEL.bracket
+    do  liftIO $ IO.forkFinally
+          do  f & runPropertIO
+                & runExceptT
+                & runResourceT
+                & flip runReaderT (PropertyEnv tvAction)
+          (liftIO . STM.atomically . STM.putTMVar tvResult)
+    do liftIO . IO.killThread
+    do \_ -> do
+        whileNothingM do
+          mActionOrResult <- liftIO $ STM.atomically do
+            mAction <- STM.tryTakeTMVar tvAction
+            case mAction of
+              Nothing -> do
+                mResult <- STM.tryTakeTMVar tvResult
+                case mResult of
+                  Just a -> pure $ Just $ Right a
+                  Nothing -> retry
+              Just action -> pure $ Just $ Left action
+
+          case mActionOrResult of
+            Nothing -> pure Nothing
+            Just (Left action) -> action >> pure Nothing
+            Just (Right (Left e)) -> liftIO $ IO.throwIO e
+            Just (Right (Right (Left e))) -> throwAssertion e
+            Just (Right (Right (Right a))) -> pure $ Just a
+
+forAll :: ()
+  => Show a
+  => Gen a
+  -> PropertyIO a
+forAll gen = do
+  PropertyEnv tmvAction <- ask @PropertyEnv
+  mvA <- liftIO STM.newEmptyTMVarIO
```

Comment:
Same question as for `liftTest` - why do we need `TMVar` here? Can `forAll` be implemented in terms of `liftTest`?